src/main/java/org/spongepowered/api/item/data/ListItemData.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * This file is part of Sponge, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered.org <http://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.item.data;
+
+import com.google.common.base.Optional;
+
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * Represents an item that can have a {@link Collection} of types of elements.
+ * Since {@link ItemData} is meant to be mutable, the {@link ListItemData}
+ * too is mutable.
+ *
+ * <p>The various implementations of {@link ListItemData} include:
+ * {@link PagedItemData}, {@link LoreItemData}, and {@link PlaceableItemData}.</p>
+ *
+ * @param <T> The type of element involved in the collection
+ * @param <E> The implementing item data that should be comparable to itself
+ */
+public interface ListItemData<E, T extends ListItemData<E, T>> extends ItemData<T> {
+
+    /**
+     * Gets all the elements stored by this {@link ListItemData}.
+     *
+     * <p>This will return an unmodifiable list of elements. To modify the
+     * list of elements, use the provided methods in {@link ListItemData}.</p>
+     *
+     * @return An unmodifiable list of all elements
+     */
+    List<E> getAll();
```

Comment:
Should the index in this list map directly onto the index of the ListItemData?
I ask mainly because it could potentially contain `null` values where there is no data at some index.